design-proposals/volume-update-strategy.md

```diff
@@ -0,0 +1,229 @@
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+This proposal introduces a new field in the VM spec to represent the volume update strategy called `updateVolumesStrategy`. This value is a unique field per VM, it can takes the values:
+ * `replacement` when the update of the volume aims to substitute the volume with another upon a VM restart
+ * `migration` when the update of the volume wants to migrate a volume to another one performing a full-copy of the content
+ 
+If no value is specified, the default behavior is `replacement`.
+ 
+Possibly, in the future, we could add a strategy for hotplugged volumes when the update wants to immediately add a new volume to a running VM.
+
+## API examples
+
+#### VM Spec API update
+
+Original VM spec
+```yaml
+  spec
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: src-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: src-pvc2
+  - name: vol3
+     persistentVolumeClaim:
+    claimName: src-pvc3
+```
+
+Update VM spec:
+```yaml
+spec:
+  updateVolumesStrategy: migration
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: dst-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: dst-pvc2
+  - name: vol3
+    persistentVolumeClaim:
+      claimName: src-pvc3
+```
+
+In this, example the user is able to update and migrate the first and second volume while the third one remains untouched.
+
+## Flow
+
+The volume migration update strategy follow the same [update sequence](https://github.com/kubevirt/community/blob/main/design-proposals/cpu-hotplug.md#phase-2-triggering-a-live-migration-when-the-number-of-sockets-changes-to-adjust-the-virt-launcher-cpu-resources) proposed for memory and cpu hotplug.
+
+Sequence of events if the `updateVolumesStrategy` is migration:
+
+* the user updates the volumes on a vm.spec.template.spec.domain.volumes
+* the VM controller determines the volumes which needs to be updated 
+* the VM controller validates if the volumes can be migrated
+* the VM controller updates the volumes accordingly in the active vmi.Spec
+* the VMI controller writes the required volumes to migrate in the vmi.Status and set the update condition in the VMI status
+* the VMI workload-update controller detects the update condition `VolumesChange` and triggers the VMI migration. When it creates the VMI migration object it also set the label `kubevirt.io/volume-update-migration: vm-name`.
+* the Migration controller constructs a new target pod with the desired volumes
+* virt-launcher updates the volumes to migrate in the migration request for LibVirt
+* the VMI live migrates
+* the VMI status is updated to reflect that the volume update action have completed and the update condition is removed
+
```

Comment:
Question @aglitke brought up offline:  What happens if `updateVolumesStrategy` is changed from `migration` while a migration is in progress?

Reply:
This correspond to 2 VM spec versions. The answer depends is if the migrated volume is also update when the `updateVolumesStrategy` is changed.

If the migrated volumes aren't changed, then the VM will still migrate and it will have a condition that the change will be effective on restarts.

If the update concerns also one of the migrated volumes, the VM is still migrating and the volumes will be copied and the VM will be marked that it requires a restart for the new update. 
However, we might encounter a conflicts if the VM is shutdown while it is still migrating. I guess in this case, we can interpret it as an abortion an simply use the latest version of the VM and don't perform the volume migration.
The original PVC won't be deleted, so there isn't any loss of data and the user can always revert the VM definition.

Does this sound reasonable to you?

Reply:
I'm not sure I understand.  Consider the following scenario:

1.  User sets `updateVolumesStrategy` to `migration`.  Removes volume A and adds volume B to the list of volumes.
2. Internally we start migrating volume A to Volume B
3. While migration is still in progress, the user edits the VM to change `updateVolumesStrategy` to `replace` and adds volume C

I assume migration from A to B continue?  C will be added when the VM reboots?

What if volume B is also removed is step 3?  Will the migration be aborted?

Reply:
> I'm not sure I understand. Consider the following scenario:
> 
> 1. User sets `updateVolumesStrategy` to `migration`.  Removes volume A and adds volume B to the list of volumes.
> 2. Internally we start migrating volume A to Volume B
> 3. While migration is still in progress, the user edits the VM to change `updateVolumesStrategy` to `replace` and adds volume C
> 
> I assume migration from A to B continue? C will be added when the VM reboots?

That's what I would expect too.

> 
> What if volume B is also removed is step 3? Will the migration be aborted?

Do we already have a way to cancel an ongoing volume migration?

Reply:
> I'm not sure I understand. Consider the following scenario:
> 
> 1. User sets `updateVolumesStrategy` to `migration`.  Removes volume A and adds volume B to the list of volumes.
> 2. Internally we start migrating volume A to Volume B
> 3. While migration is still in progress, the user edits the VM to change `updateVolumesStrategy` to `replace` and adds volume C
> 
> I assume migration from A to B continue? C will be added when the VM reboots?

Yes, the migration still continues and yes C will be added when the VM reboots
> 
> What if volume B is also removed is step 3? Will the migration be aborted?

The migration should be aborted only if we reboot the VM during the migration, otherwise, this should continue.

Is it  clearer?

Reply:
@mhenriks @vladikr @acardace after I started to implement the abortion part, I realized that allow any volume change while the VM is migrating can lead to a very confusing state. Therefore, I decided that we allow a volume change while we are migrating only if we revert the volumes to the previous one. Basically, changing the destination volumes back to the source ones. 
For cpu and memory hotplug, we return an error if the VMI is still migrating. Hence, this follow more or less the same pattern with the difference that the migration can be aborted if we return to the previous volume set.
Does this sound reasonable to you?

Reply:
@alicefr yes, it sounds reasonable.

Reply:
> we allow a volume change while we are migrating only if we revert the volumes to the previous one.

Say a VM has two volume A and B.  We start migrating A to C.  While that is in progress we decide to migrate B to D.  I assume that change will be blocked?

Reply:
Yes, while the update is in progress you can only change C -> A while C ->D will be rejected

Reply:
Changing C->A will abort the migration.